armv7m7: Fix syspage arguments parsing

gerard5 opened the pull request:
This change fixes parsing of arguments passed to syspage and error checking in plo app command. Prior to this change arguments were truncated e.g `app flash0 @lwip;pppou:/dev/uart3:460800:up xip1 ocram2` was resolved as process name `lwip` with argv[1]= `pppou:/dev/` but expected was argv[1]=`pppou:/dev/uart3:460800:up`.

Working examples based on correct syntax:
```
app flash0 @lwip;pppou:/dev/uart3:460800:up xip1
app flash0 @ps_g3_mac_srv;msg=/dev/sap1;band=cen_a;run=3 xip1 ocram2
```
and downloading app to internal ram through auxiliary uart (using phoenixd):
```
app com1 g3adp_test;log7;m=/dev/sap1;testcase=2:8:0:18
```
JIRA: RTOS-4 , RTOS-5

kemonats commented on armv7m7-imxrt106x/syspage.c:
```diff
@@ -23,6 +23,7 @@
 #define MAX_ENTRIES_NB          6     /* 3 of kernel's sections, 2 of plo's sections and syspage */
+#define MAX_CMDLINE_SIZE        16
 
 #define MAX_SYSPAGE_SIZE     (sizeof(syspage_t) + MAX_ARGS_SIZE * sizeof(char) + MAX_PROGRAMS_NB * sizeof(syspage_program_t) + MAX_MAPS_NB * sizeof(syspage_map_t))
```
The standard `C` probably guarantees `sizeof(char) == 1`.

gerard5 replied:
I see, but this does not apply to topic of this change (PR).

kemonats commented on armv7m7-imxrt106x/syspage.c:
```diff
@@ -436,35 +439,48 @@ void syspage_addEntries(u32 start, u32 sz)
 
+	len = plostd_strlen(name);
+
+	if (syspage_common.argCnt + 1 + len + 1 > MAX_ARGS_SIZE) {
```
```suggestion
	if (syspage_common.argCnt + 1 + len + 1 + 1 > MAX_ARGS_SIZE) {
```
`'X' + '  ' + '\0'`

gerard5 replied:
You're right, corrected.

kemonats commented on armv7m7-imxrt106x/syspage.c:
```diff
@@ -436,35 +439,48 @@ void syspage_addEntries(u32 start, u32 sz)
+		plostd_printf(ATTR_ERROR, "\nSyspage program %s, name is too long!\n", name);
+		return -1;
+	}
+
```
What if `pos == MAX_CMDLINE_SIZE`.
`low_memcpy()` copy `pos chars`.
and 
```
	while (pos < MAX_CMDLINE_SIZE)
		syspage_common.progs[progID].cmdline[pos++] = 0;
```
 not add `\0` at end.

gerard5 replied:
Thank You, corrected.

kemonats commented on armv7m7-imxrt117x/syspage.c:
```diff
@@ -23,6 +23,7 @@
 #define MAX_ENTRIES_NB          6     /* 3 of kernel's sections, 2 of plo's sections and syspage */
+#define MAX_CMDLINE_SIZE        16
 
 #define MAX_SYSPAGE_SIZE     (sizeof(syspage_t) + MAX_ARGS_SIZE * sizeof(char) + MAX_PROGRAMS_NB * sizeof(syspage_program_t) + MAX_MAPS_NB * sizeof(syspage_map_t))
```
Same as for `imxrt106x`.

gerard5 replied:
See the answer for imxrt106x.

kemonats commented on armv7m7-imxrt117x/syspage.c:
```diff
@@ -438,35 +439,48 @@ void syspage_addEntries(u32 start, u32 sz)
 
+	len = plostd_strlen(name);
+
+	if (syspage_common.argCnt + 1 + len + 1 > MAX_ARGS_SIZE) {
```
Same as for `imxrt106x`.

gerard5 replied:
As for imxrt106x, updated.

kemonats commented on armv7m7-imxrt117x/syspage.c:
```diff
@@ -438,35 +439,48 @@ void syspage_addEntries(u32 start, u32 sz)
+		plostd_printf(ATTR_ERROR, "\nSyspage program %s, name is too long!\n", name);
+		return -1;
+	}
+
```
Same as for `imxrt106x`.

gerard5 replied:
As for imxrt106x, updated.

kemonats commented on armv7m7-imxrt106x/syspage.c:
```diff
@@ -436,35 +439,48 @@ void syspage_addEntries(u32 start, u32 sz)
 
+	len = plostd_strlen(name);
+
+	if (syspage_common.argCnt + 1 + len + 1 + 1 > MAX_ARGS_SIZE) {
```
Sorry, but I think I may have made a mistake.
The `argsCnt` points to `\0'`so you need an extra 2 bytes of `'X'` + `' '` not 3.
Check it again.

gerard5 replied:
If you mean `argCnt +  len + 2 >= MAX_ARGS_SIZE`  than `argCnt + len + 3 > MAX_ARGS_SIZE` is the equivalent, but I prefer the current version as it is more readable `X` + `len` + ` ` + `\0`.
Please see the attached figure, for an explanation:

![qw2e59](https://user-images.githubusercontent.com/141153/110604442-6251de80-8188-11eb-967f-cb63a55cc681.png)

MaciejPurski commented:
hw tested this PR on imxrt1176x. Works as expected.

agkaminski reviewed (approved):
LGTM

gerard5 commented:
@HBuczynski , because Your PR introducing Zynq target was recently merged, I rebased and updated my PR to Zynq target (due to common code shared between `imxrt106x`, `imxrt117x` and `zynq` - the same problem with truncating syspage long arguments also applies to the Zynq target).

@HBuczynski, please review the PR code (at least the zynq part) and verify on hardware before merging.

HBuczynski commented:
works correctly on zynq
